File: 알고리즘(Algorithms)/1. 정렬(Sorting)/Quick sort/Quick sort.c
#include <stdio.h>

void swap(int arr[], int a, int b)
{
	int temp = arr[a];
	arr[a] = arr[b];
	arr[b] = temp;
}

int Partition(int arr[], int from, int to)
{
	int pivot = arr[to];
	int i = from - 1;
	for (int j = from; j <= to - 1; j++) {
		if (arr[j] <= pivot) {
			i++;
			swap(arr, i, j);
		}
	}
	swap(arr, i + 1, to);
	return i + 1;
}

void quickSort(int arr[], int from, int to)
{
	if (from < to) {
		int partition = Partition(arr, from, to);
		quickSort(arr, from, partition - 1);
		quickSort(arr, partition + 1, to);
	}
}

int main(void)
{
	int arr[10] = { 2,4,5,3,9,1,0,2,3,7 };
	quickSort(arr, 0, 9);
	for (int i = 0; i < 10; i++) printf("%d ", arr[i]);
	return 0;
}

File: 알고리즘(Algorithms)/1. 정렬(Sorting)/Quick sort/Quick sort.java

public class QuickSort {
	static void swap(int arr[], int a, int b) {
		int temp=arr[a];
		arr[a]=arr[b];
		arr[b]=temp;
	}
	
	static void quickSort(int arr[], int from, int to) {
		if(from<to) {
			int m = Partition(arr,from,to);
			quickSort(arr,from,m-1);
			quickSort(arr,m+1,to);
		}
	}
	static int Partition(int arr[], int from, int to) {
		int p = arr[from];
		int i=from, j=from+1;
		while(j<=to) {
			if(arr[j]<=p) {
				i++;
				swap(arr,i,j);
			}
			j++;
		}
		swap(arr, from ,i);
		return i;
	}
	
	public static void main(String[] args) {
		int arr[]= {4,3,6,2,1,9,5,4,0,2,3};
		quickSort(arr,0,10);
		for(int i=0; i<=10; i++) System.out.println(arr[i]);
	}
}
